### ISSUE TITLE
grad fails through jax.numpy.square and jax.numpy.linalg.norm at zero

### ISSUE THREAD

**Issue by NeilGirdhar** [google/jax#2377]
Here's a minimum example:
```
import jax.numpy as jnp
from jax import grad


x = 0.0
target = 1.0


def loss1(w):
    y = jnp.sqrt(((x * w) ** 2).sum())
    return (y - target)**2


def loss2(w):
    y = jnp.sqrt(jnp.square(x * w).sum())
    return (y - target)**2


def loss3(w):
    y = jnp.linalg.norm(x * w)
    return (y - target)**2


for loss in [loss1, loss2, loss3]:
    print(grad(loss)(1.0))
```
Prints:
```
0.0
nan
nan
```

**Comment by mattjj** [comment]
A blast from the past! See hips/autograd#151 and [this comment](https://github.com/google/jax/issues/1052#issuecomment-514083352) on #1052. I think these should probably all be nan.

What answer did you expect?

**Comment by NeilGirdhar** [comment]
I'm sorry, I don't understand how your comment applies?  Where is the nan arising?

**Comment by NeilGirdhar** [comment]
Oh, I see because they're all like the derivative of `abs`.  I would still prefer to arbitrarily choose zero as the derivative.  There must be some nice way of having that happen.  Like defining the derivative as the average of directional derivatives?  Or even better letting me take a directional derivative, and I'll take the average over the directions I want.  Is that possible?

**Comment by NeilGirdhar** [comment]
Sorry for all the edits in my replies.  You are right.  Choose whatever is easiest.

**Comment by mattjj** [comment]
I only just checked, so I didn't see the edits :)

Yes, I should have explained: they're all like ways of writing `abs`, and so the corresponding mathematical function is not differentiable at zero.

However, just like we do for `np.abs`, we could choose a convention for a reasonable non-nan value to produce at zero (like an element of the subgradient, which often but not always makes sense in the context of gradient-based optimization). That's what you suggested too: choose an element of the convex hull of directional derivatives. (I'm not sure of how to provide that choice to the user. Also I've been meaning to read [this paper](https://arxiv.org/abs/1809.08530) but I haven't yet.)

But there's another subtlety: different ways of denoting a mathematical function as a program (e.g. expressing `abs = lambda x: np.abs(x)` versus `abs = lambda x: np.sqrt(x**2)`) can, after applying autodiff, result in derivative programs that denote different mathematical functions! A primary goal of autodiff is to make them the same, since what does autodiff mean if not "produce a program that evaluates the derivative of the mathematical function denoted by the input program", but sometimes we fail. (Here's an extreme example Martín Abadi showed us: `f = lambda x: -x if x == 0 else x`. That's a denotation of the identity function, so there's no mathematical pathology: it's infinitely differentiable. Yet in most autodiff systems `grad(f)(0.) == -1`! We built JAX to be able to detect this and raise an error, though we didn't switch it on because no one seems to care much, including us. It's nice to know about though!)

We'd been working on autodiff pretty much nonstop for 4+ years by the time we noticed the subtler points of these issues :) And even now there are some mistakes in JAX (as per my comment in #1052, `safe_mul` was a mistake). We should identify and clean up those mistakes, though my current starting point is that nans might be the right thing in some of these cases.

Hope that helps! Thanks for raising this.
